game-core/src/main/java/games/strategy/triplea/delegate/battle/AbstractBattle.java

```diff
@@ -97,6 +98,14 @@
     return Collections.unmodifiableList(dependentUnits);
   }
 
+  protected Collection<Unit> getWithDependents(final Collection<Unit> units) {
+    final Collection<Unit> dependentUnits = getDependentUnits(units);
+    if (dependentUnits.isEmpty()) {
+      return units;
+    }
+    return ImmutableList.copyOf(Iterables.concat(units, dependentUnits));
```

Comment:
Does `concat` of a non-empty collection with an empty collection return just the non-empty collection? Perhaps we should just let the API do that for us?

In, part, it's strange that if you pass an mutable list to `getWithDependents` that is empty, you get a mutable list back from this method, but passing a non-empty mutable list returns an immutable collection. Generally a returned collection should always be mutable or immutable and not depend on its parameter.

Reply:
`getWithDependents` is notable to me as a method name. It makes me wonder 'get what?'

`getUnitWithDependents` makes perhaps more sense. Though, if we have a one-liner method, perhaps just in-lining this in the two places its used is just as well? WDYT?

Reply:
> Does `concat` of a non-empty collection with an empty collection return just the non-empty collection? Perhaps we should just let the API do that for us?

It returns an iterable. Creating a list of it does require a copy.

> 
> In, part, it's strange that if you pass an mutable list to `getWithDependents` that is empty, you get a mutable list back from this method, but passing a non-empty mutable list returns an immutable collection. Generally a returned collection should always be mutable or immutable and not depend on its parameter.

We can wrap units in an immutable wrapper for consistency.

Reply:
Renamed to `getUnitsWithDependents` and made the return value unmodifiable.

Reply:
Sorry to not have been clear in my question:

When calling:
```ImmutableList.copyOf(Iterables.concat(units, dependentUnits))```

Does it matter if `dependentUnits` is empty? Wouldn't we still get the same result?

Reply:
Side-note, I'm not sure we should be completely zeolous to wrap collections in unmodifiable variants. IMO it's far more important for the contents to be immutable than a collection.

We were concerned about performance of unnecessary collection copies, seems we would be falling into that potential problem more and more. One thing to do is for us to simply stop mutating collections though and focus on instead of copying collections, to instead do stream operations on them which creates a new collection for us. This way we avoid both mutating collections and copying them explicitly as part of return or input arguments.